### ISSUE TITLE
[question] resolving header name conflicts among different dependencies

### ISSUE THREAD

**Issue by FWiesner** [conan-io/conan#17222]
### What is your question?

Hi @memsharded. We are having the problem that we have package header names clashing between a system package and packages built with recipes from the Conan index. I was thinking to modify the respective Conan Center recipe to add an option for the header path to be modified to allow `#include <prefix/sql.h>` instead of `#include <sql.h>`, but doing so in the recipe feels awkward and if taken to the extreme, would require a lot of recipe adjustments.

Therefore, I wonder whether there already is a feature in 2.8.1 or the conan-cmake integration that could be used to modify the include path on consumption to achieve the same without recipe hacking.

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @FWiesner 

Thanks for your question

> We are having the problem that we have package header names clashing between a system package and packages built with recipes from the Conan index. I was thinking to modify the respective Conan Center recipe to add an option for the header path to be modified to allow #include <prefix/sql.h> instead of #include <sql.h>, but doing so in the recipe feels awkward and if taken to the extreme, would require a lot of recipe adjustments.

I am afraid that this is complicated, it can be done of course, but it is "viral" it might require changes in all the downstream consumers. It is indeed a good practice for C++ to always use some prefix in headers, but many open source libraries don't do it. So I agree it is awkward, I think I would discourage against it.

Conan can't do much here really, besides what it is already doing. Conan prioritizes the include headers paths in the build systems, so Conan packages should be found instead of potential system alternatives, and there shouldn't be any conflict in this sense. If there is some cases in which the system package is found instead of the Conan one, that probably needs to be investigated and checked.

**Comment by FWiesner** [comment]
>  it might require changes in all the downstream consumers

that would only be the case if the header path override option in the recipe would be set by default or when the ~build~ host profile would have been updated to set the override and then another downstream package would expect the headers in the standard location, right?

**Comment by FWiesner** [comment]
btw @memsharded... after setting

```python
if self.options.includes_prefix is not None:
  self.cpp.package.includedirs = [f"include/{self.options.includes_prefix}"]
```

in the `layout` function, I see that the built bundle (I've modified the `odbc` recipe) directory contains the headers under `.conan2/p/b/odbcf09be5958dde7/p/include/prefix`, but they won't be added to the using targets' CMake include directories.

adding

```python
if self.options.includes_prefix != None:
   self.cpp_info.includedirs = ["include"]
```

in `package_info` makes no difference there

**Comment by memsharded** [comment]
> that would only be the case if the header path override option in the recipe would be set by default or when the build host profile would have been updated to set the override and then another downstream package would expect the headers in the standard location, right?

I am not sure I follow. The only way to introduce such a prefix, is path-based, that is:

- The ``package()`` method of the recipe must be changed, so headers are copied to ``include/prefix`` instead of copied to ``include``
- The ``cpp_info.includedirs = ["include"]`` remains.
- It is not possible to make something in the recipe level that allows a ``#include <prefix/header.h>`` without actually having the ``header.h`` inside a folder called ``prefix``, this is necessary.
- The moment the header locations are changed, there are 2 options:
   - All consumers must use ``#include <prefix/header.h>``. That means, for open source packages that it is necessary to patch the source code of all consumers that almost surely will depend on the standard ``#include <header.h>``
   - If the recipe adds ``cpp_info.includedirs = ["include/prefix"]`` and ``#include <header.h>`` is allowed, we are back to square one, and the system header can be found again.


> if self.options.includes_prefix is not None:

Be careful, this is not valid Conan recipe code. You need to use ``if not self.options.includes_prefix`` or ``if self.options.includes_prefix != None``, as the ``includes_prefix`` is not a primitive type


> self.cpp.package.includedirs = [f"include/{self.options.includes_prefix}"]

This is what I described above, if you allow ``#include <header.h>``, then nothing has changed, and it is possible that consumers would find the system one.

Still, it is not very clear what is the issue, as described above, Conan generators should prioritize the Conan package paths, so even if the headers are in the system, the consumer should find the Conan package headers, not the system headers.

**Comment by FWiesner** [comment]
I agree on the potential conflicts that such an option can bring. But, I assume this conflict would only be a problem when the option was set by a project that itself would have a recipe on Conan Center. But, when set by a project that itself is not exported and when the dependency is not appearing as transitive one, I don't think it would be a problem.

Probably an additional optional qualifier on options that would "taint" derived project Conan recipes, could be a solution. When a recipe is tainted, you either would not be able to require it in other conanfiles or you get warnings of a dangerous setup.

**Comment by memsharded** [comment]
I am not sure if I understood. The options don't bring new potential conflicts. It just leave the problem exactly as it were without the option, because things will still resolve using the include-paths priority. So I still don't understand how an option could help.

There are mechanisms like the ``[platform_requires]`` or using ``[replace_requires]`` with a ``pkg/system`` recipe that allows to decide from the downstream if the Conan package transitive dependency is used or the system one, I am not sure if this is something you could be looking for.

So I am afraid it is not clear what the issue or problem to solve is. We would need a reproducible case, a minimal project or something like that that demonstrate how a build will use the system headers instead of the Conan package headers, we haven't understood yet the initially reported problem of the headers.

**Comment by FWiesner** [comment]
our specific problem is that there is a commercial uber-package in our case that must be first in the hunting chain due to a custom C runtime, but it comes with an outdated version of ODBC that cannot be removed easily as there are tons of local includes in the headers of that package and all headers are in the same directory. So, we need a second ODBC header bundle that is loadable after the commercial package and for that we need the prefix.

The conflict I was talking about would be when there would be say a library that would also look for ODBC headers, but would get the old ones, as headers like sql.h would be loaded from the un-prefixed location

**Comment by memsharded** [comment]
This is what I mean, it is a bit difficult to understand this from the outside, without a little bit more of context.
If not having things expressed as code, we'd appreciate something like:
- We have a Conan package named ``mypkg/1.0`` that we create from a commercial library. 
- This library contains its headers in the package folder n the ``include/mypkg/myheader.h``.
- Consumers of this package will use it with a normal ``requires="mypkg/1.0", and they will ``#include "mypkg/myheader.h``. 
- The consumers are using CMakeDeps and CMakeToolchain.
- ...

Because for example, it is not clear if the ``a commercial uber-package`` is a Conan package or not, etc.

Thanks for your feedback!

**Comment by memsharded** [comment]
Hi @FWiesner 

I'd like to follow up on this. Is there any further detail or feedback here?
Did you solve the issues with those headers? Thanks for the feedback.

**Comment by memsharded** [comment]
Closing this ticket as responded and staled. If you have any further questions or feedback, please comment, re-open or create a new ticket. Thanks for your feedback!
